src/Compilers/CSharp/Portable/Symbols/Source/ExtensionGroupingInfo.cs

```diff
@@ -0,0 +1,504 @@
+        }
+
+        private sealed class ExtensionGroupingType : ExtensionGroupingOrMarkerType, IComparable<ExtensionGroupingType>
+        {
+            private readonly string _name;
+            public readonly ImmutableArray<ExtensionMarkerType> ExtensionMarkerTypes;
+            private readonly ImmutableArray<ExtensionGroupingTypeParameter> _typeParameters;
+
+            public ExtensionGroupingType(string name, MultiDictionary<string, SourceNamedTypeSymbol> extensionMarkerTypes)
+            {
+                _name = name;
+
+                var builder = ArrayBuilder<ExtensionMarkerType>.GetInstance(extensionMarkerTypes.Count);
+
+                foreach (var pair in extensionMarkerTypes)
+                {
+                    builder.Add(new ExtensionMarkerType(this, pair.Key, pair.Value));
+                }
+
+                builder.Sort();
+                ExtensionMarkerTypes = builder.ToImmutableAndFree();
+
+                _typeParameters = ExtensionMarkerTypes[0].UnderlyingExtensions[0].Arity != 0 ?
+                    ((INestedTypeDefinition)ExtensionMarkerTypes[0].UnderlyingExtensions[0].GetCciAdapter()).GenericParameters.SelectAsArray(static (p, @this) => new ExtensionGroupingTypeParameter(@this, p), this) :
+                    [];
+            }
+
+            int IComparable<ExtensionGroupingType>.CompareTo(ExtensionGroupingType? other)
+            {
+                Debug.Assert(other is { });
+                return ExtensionMarkerTypes[0].CompareTo(other.ExtensionMarkerTypes[0]);
+            }
+
+            protected override IEnumerable<IGenericTypeParameter> GenericParameters => _typeParameters;
+
+            protected override ushort GenericParameterCount => (ushort)ExtensionMarkerTypes[0].UnderlyingExtensions[0].Arity;
+
+            protected override bool IsAbstract => false;
+
+            protected override bool IsSealed => true;
+
+            protected override ITypeDefinition ContainingTypeDefinition => ExtensionMarkerTypes[0].UnderlyingExtensions[0].ContainingType!.GetCciAdapter();
+
+            protected override TypeMemberVisibility Visibility => TypeMemberVisibility.Public;
+
+            public override string Name => _name;
+
+            protected override ITypeReference? ObjectType => ExtensionMarkerTypes[0].UnderlyingExtensions[0].ContainingAssembly.GetSpecialType(SpecialType.System_Object).GetCciAdapter();
+
+            protected override IEnumerable<IMethodDefinition> GetMethods(EmitContext context)
+            {
+                foreach (var marker in ExtensionMarkerTypes)
+                {
+                    foreach (var type in marker.UnderlyingExtensions)
+                    {
+                        foreach (var method in type.GetMethodsToEmit())
+                        {
+                            Debug.Assert((object)method != null);
+
+                            // PROTOTYPE: SynthesizedExtensionMarker should not be added as a member of the extension type.
```

Comment:
I'm not clear what you intend for `SynthesizedExtensionMarker`. Consider leaving more details if this is something I'll need to handle